authorization/storage.go

```diff
@@ -2,33 +2,313 @@ package authorization
-		IDGen:   snowflake.NewDefaultIDGenerator(),
+type storePlusOptions struct {
+	*Store
+	hasherVariantName string
+}
+
+type StoreOption func(*storePlusOptions)
+
+func WithAuthorizationHasher(hasher *AuthorizationHasher) StoreOption {
+	return func(s *storePlusOptions) {
+		s.hasher = hasher
+	}
+}
+
+func WithAuthorizationHashVariantName(name string) StoreOption {
+	return func(s *storePlusOptions) {
+		s.hasherVariantName = name
+	}
+}
+
+func WithIgnoreMissingHashIndex(allowMissing bool) StoreOption {
+	return func(s *storePlusOptions) {
+		s.ignoreMissingHashIndex = allowMissing
+	}
+}
+
+func WithLogger(log *zap.Logger) StoreOption {
+	return func(s *storePlusOptions) {
+		s.log = log
+	}
+}
+
+func WithReadOnly(readOnly bool) StoreOption {
+	return func(s *storePlusOptions) {
+		s.readOnly = readOnly
 	}
-	return st, st.setup()
+}
+
+// NewStore creates a new authorization.Store object. kvStore is the underlying key-value store.
+func NewStore(ctx context.Context, kvStore kv.Store, useHashedTokens bool, opts ...StoreOption) (*Store, error) {
+	s := &storePlusOptions{
+		Store: &Store{
+			kvStore:         kvStore,
+			IDGen:           snowflake.NewDefaultIDGenerator(),
+			useHashedTokens: useHashedTokens,
+		},
+		hasherVariantName: DefaultHashVariantName,
+	}
+
+	for _, o := range opts {
+		o(s)
+	}
+
+	if s.log == nil {
+		s.log = zap.NewNop()
+	}
+
+	if err := s.setup(ctx); err != nil {
+		return nil, err
+	}
+
+	if s.hasher == nil {
+		hasher, err := s.autogenerateHasher(ctx, s.hasherVariantName)
+		if err != nil {
+			return nil, err
+		}
+		s.hasher = hasher
+	}
+
+	// Perform hashed token migration if needed. This can not be performed by the migration service
+	// because it requires configuration, and the migration service is more concerned with schema
+	// and does not have configuration.
+	if err := s.hashedTokenMigration(ctx); err != nil {
+		return nil, fmt.Errorf("error during hashed token migration: %w", err)
+	}
+
+	return s.Store, nil
+}
+
+// autogenerateHasher generates an AuthorizationHasher that hashes using variantName.
+// The decoders include variantName plus any other variants that are detected in the
+// store.
+func (s *Store) autogenerateHasher(ctx context.Context, variantName string) (*AuthorizationHasher, error) {
+	// Determine which variants are present in the store.
+	tempDecoder := crypt.NewDecoder()
+	if err := influxdb2_algo.RegisterDecoder(tempDecoder); err != nil {
+		return nil, fmt.Errorf("error registering test decoders for authorization store: %w", err)
+	}
+
+	var auths []*influxdb.Authorization
+	err := s.View(ctx, func(tx kv.Tx) error {
+		as, err := s.ListAuthorizations(ctx, tx, influxdb.AuthorizationFilter{})
+		if err != nil {
+			return err
+		}
+		auths = as
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error fetching authorization records for hash variant inventory: %w", err)
+	}
+
+	foundVariants := make(map[influxdb2_algo.Variant]struct{})
+	for _, a := range auths {
```

Comment:
Since auth is a pointer should we add a nil check prior to looping over it as a safety check.

Reply:
`auths` is a slice of pointers, not a pointer itself.